File: backend/app/schemas/generation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class GenerationRequest(BaseModel):
    prompt_id: UUID
    num_samples: int = Field(1, ge=1, le=4)
    adapter_id: UUID | None = None
    seed: int | None = None
    temperature: float = Field(1.0, ge=0.1, le=2.0)
    top_k: int = Field(250, ge=1, le=1000)
    top_p: float = Field(0.0, ge=0.0, le=1.0)
    duration: int | None = Field(None, ge=1)  # No max limit, user chooses freely


class GenerationJobResponse(BaseModel):
    id: UUID
    status: str
    progress: float | None = None
    audio_ids: list[UUID] | None = None
    error: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


# Job Feedback Schemas
class SampleFeedbackItem(BaseModel):
    """Individual feedback record for a sample."""

    id: UUID
    rating: float | None
    rating_criterion: str | None
    preferred_over: UUID | None
    tags: list[str] | None
    notes: str | None
    created_at: datetime

    class Config:
        from_attributes = True


class SampleFeedbackGroup(BaseModel):
    """Feedback grouped by audio sample."""

    audio_id: UUID
    label: str  # A, B, C, D...
    feedback: list[SampleFeedbackItem]
    average_rating: float | None = None
    feedback_count: int = 0
    tags: list[str] | None = None  # Audio tags from AudioTag table


class JobFeedbackResponse(BaseModel):
    """Response for GET /generate/{job_id}/feedback."""

    job_id: UUID
    prompt_id: UUID
    total_samples: int
    total_feedback: int
    average_rating: float | None
    samples: list[SampleFeedbackGroup]

File: frontend/src/lib/api/types/generation.ts
/**
 * Generation job-related types
 */

export interface GenerationJob {
  id: string;
  status: 'queued' | 'processing' | 'completed' | 'failed';
  progress?: number;
  audio_ids?: string[];
  error?: string;
}

export interface JobDetail {
  id: string;
  prompt_id: string | null;
  prompt_preview: string | null;
  adapter_id: string | null;
  adapter_name: string | null;
  status: string;
  progress: number;
  num_samples: number;
  audio_ids: string[];
  error: string | null;
  duration_seconds: number | null;
  created_at: string;
  completed_at: string | null;
}

export interface JobStats {
  status_counts: Record<string, number>;
  active_jobs: number;
  avg_processing_time_seconds: number | null;
  jobs_today: number;
  total_jobs: number;
}

export interface SubmitGenerationRequest {
  prompt_id: string;
  num_samples?: number;
  adapter_id?: string;
  duration?: number;  // Audio duration in seconds (1-120)
}

// Job Feedback Types
export interface SampleFeedbackItem {
  id: string;
  rating: number | null;
  rating_criterion: string | null;
  preferred_over: string | null;
  tags: string[] | null;
  notes: string | null;
  created_at: string;
}

export interface SampleFeedbackGroup {
  audio_id: string;
  label: string;
  feedback: SampleFeedbackItem[];
  average_rating: number | null;
  feedback_count: number;
  tags: string[] | null;
}

export interface JobFeedbackResponse {
  job_id: string;
  prompt_id: string;
  total_samples: number;
  total_feedback: number;
  average_rating: number | null;
  samples: SampleFeedbackGroup[];
}
